packages/client/src/blob/read_progress.rs

```diff
@@ -0,0 +1,85 @@
+use std::sync::atomic::AtomicU64;
+
+use crate as tg;
+use crate::blob::Reader;
+use std::pin::Pin;
+use std::sync::{atomic::Ordering, Arc};
+use tokio::io::{AsyncBufRead, AsyncRead, AsyncSeek};
+
+pub struct ProgressReader<H> {
+	inner: Reader<H>,
+	position: Arc<AtomicU64>,
+}
+
+impl<H> ProgressReader<H>
+where
+	H: tg::Handle,
+{
+	pub fn new(inner: Reader<H>) -> tg::Result<ProgressReader<H>> {
+		Ok(ProgressReader {
+			inner,
+			position: Arc::new(AtomicU64::new(0)),
+		})
+	}
+
+	pub fn position(&self) -> Arc<AtomicU64> {
+		self.position.clone()
+	}
+
+	pub fn size(&self) -> u64 {
+		self.inner.size()
+	}
+}
+
+impl<H> AsyncRead for ProgressReader<H>
+where
+	H: tg::Handle,
+{
+	fn poll_read(
+		self: Pin<&mut Self>,
+		cx: &mut std::task::Context<'_>,
+		buf: &mut tokio::io::ReadBuf<'_>,
+	) -> std::task::Poll<std::io::Result<()>> {
+		let this = self.get_mut();
+		let poll_result = Pin::new(&mut this.inner).poll_read(cx, buf);
+		if let std::task::Poll::Ready(Ok(())) = &poll_result {
+			let read_bytes = buf.filled().len() as u64;
```

Comment:
This is ugly because we also keep track of progress in `consume()` below. In practice, we always use one or the other in extract/(de)compress. Taking suggestions on how to improve this.